language/bytecode-verifier/src/borrow_graph.rs

```diff
@@ -0,0 +1,245 @@
+    pub fn remove_nonce(&mut self, nonce: Nonce) {
+        checked_assume!(self.invariant());
+        let nonce_edge_set = self.0.remove(&nonce).unwrap();
+        let mut removed_edges = BTreeMap::new();
+        for (n, es) in &self.0 {
+            removed_edges.insert(
+                n.clone(),
+                es.iter()
+                    .filter(|x| x.to == nonce)
+                    .cloned()
+                    .collect::<Vec<_>>(),
+            );
+        }
+        for (n, es) in &removed_edges {
+            es.iter().for_each(|removed_edge| {
+                let n_edge_set_ref = self.0.get_mut(n).unwrap();
+                n_edge_set_ref.remove(removed_edge);
+                nonce_edge_set.iter().for_each(|nonce_edge| {
+                    if removed_edge.edge_type == EdgeType::Strong {
+                        let mut new_label = vec![];
+                        new_label.append(&mut removed_edge.label.clone());
+                        new_label.append(&mut nonce_edge.label.clone());
+                        let edge = Edge {
+                            edge_type: nonce_edge.edge_type,
+                            label: new_label,
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    } else {
+                        let edge = Edge {
+                            edge_type: EdgeType::Weak,
+                            label: removed_edge.label.clone(),
+                            to: nonce_edge.to,
+                        };
+                        n_edge_set_ref.insert(edge);
+                    }
+                });
+            });
+        }
+        checked_verify!(self.invariant());
+        checked_postcondition!(!self.0.contains_key(&nonce));
+    }
+
+    /// renames nonces in `self` according to `nonce_map`
+    pub fn rename_nonces(&self, nonce_map: BTreeMap<Nonce, Nonce>) -> Self {
+        checked_assume!(self.invariant());
+        let mut new_graph = BTreeMap::new();
+        for (n, es) in &self.0 {
+            new_graph.insert(
+                nonce_map[n],
+                es.iter()
+                    .map(|x| Edge {
+                        edge_type: x.edge_type,
+                        label: x.label.clone(),
+                        to: nonce_map[&x.to],
+                    })
+                    .collect::<BTreeSet<_>>(),
+            );
+        }
+        let new_borrow_graph = BorrowGraph(new_graph);
+        checked_verify!(self.invariant());
+        checked_verify!(new_borrow_graph.invariant());
+        new_borrow_graph
+    }
+
+    /// checks if `self` covers `other`
+    pub fn abstracts(&self, other: &Self) -> bool {
+        self.unmatched_edges(other).values().all(|es| es.is_empty())
+    }
+
+    /// joins `other` into `self`
+    pub fn join(&mut self, other: &Self) {
+        for (n, es) in self.unmatched_edges(other) {
+            es.into_iter().for_each(|e| {
```

Comment:
`BtreeSet` implements `Extend` [[1](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#impl-Extend%3C%26%27a%20T%3E)] and `IntoIterator` (for the operand) so you shouldn't need the iterator.

Reply:
So 
```
for (n, es) in self.unmatched_edges(other) {
    self.0.get_mut(&n).unwrap().extend(es)
}
```
?

Reply:
Done exactly like @tnowacki suggested?